File: day10-binary-search/binsearch.c
#include <stdlib.h>
#include "binsearch.h"

int binary_search(int *list, int size, int item) {
  // Returns the position of the item in the array,
  // or -1 if the item isn't present.
  int result = -1;
  int mid;
  int lower = 0;
  int upper = size - 1;

  if(list != NULL) {  // Don't be caught with your pants down.
    while(lower <= upper) {
      // Find new mid.
      mid = (lower + upper) / 2;  // This can overflow. Watch out.

      if(list[mid] < item)
	lower = mid + 1;
      else if(list[mid] > item)
	upper = mid - 1;
      else {  // list[mid] == item
	result = mid;
	break;
      }
    }
  }

  return result;
}

File: day10-binary-search/bin_search.py
def binary_search(items, target):
    '''Quickly search for an item in a list.'''
    result = -1  # Assume failure.
    lower = 0
    upper = len(items) - 1
    while lower <= upper:
        mid = (lower + upper) // 2  # Calc new mid.
        if items[mid] < target:
            lower = mid + 1
        elif items[mid] > target:
            upper = mid - 1
        else:  # items[mid] == target
            result = mid
            break
    return result
